### ISSUE TITLE
No way to know whether a information recipient is primary or copy

### ISSUE THREAD

**Issue by Kusnaditjung** [nhsconnect/integration-adaptor-111#120]
**Description**
The original CDA has two types of information recipient, primary (typecode=PRCP) or copy (typecode=TRC) recipient. The UI rendering part which uses the NHS111 adapter requires the ability to differentiate between the two, but at the moment the FHIR transform does not seem to output this flag accordingly. 

**Component**
NHS111 adapter

**Version**
0.1.0

**Step to Reproduce**
* Put the valid a CDA in the payload, and send
* Add another information recipient of type copy recipient typecode="TRC"

```xml

   <informationRecipient typeCode="TRC">
                            <npfitlc:contentId root="2.16.840.1.113883.2.1.3.2.4.18.16"
                                               extension="COCD_TP145203GB03#IntendedRecipient"/>
                            <intendedRecipient classCode="ASSIGNED">
                                <templateId extension="COCD_TP145203GB03#IntendedRecipient"
                                            root="2.16.840.1.113883.2.1.3.2.4.18.2"/>
                                <addr use="WP">
                                    <streetAddressLine>Virtual Hospital</streetAddressLine>
                                    <streetAddressLine>Adam Street</streetAddressLine>
                                    <streetAddressLine>Medway</streetAddressLine>
                                    <streetAddressLine>Essex</streetAddressLine>
                                    <postalCode>ME4 5TT</postalCode>
                                </addr>
                                <telecom use="WP" value="tel:0123476895"/>
                                <receivedOrganization classCode="ORG" determinerCode="INSTANCE">
                                    <templateId root="2.16.840.1.113883.2.1.3.2.4.18.2"
                                                extension="COCD_TP145203GB03#representedOrganization"/>
                                    <id root="2.16.840.1.113883.2.1.3.2.4.19.2" extension="PRA123"/>
                                    <name>Virtual Hospital</name>
                                </receivedOrganization>
                            </intendedRecipient>
                        </informationRecipient>
```

**Test Data**
Any data

**Preconditions**
None

**Environment**
local environment

**Severity**
Low

**Priority**
medium, may impact integration

**Expected Behavior**
The information recipient in FHIR (which becomes HealthcareService) can be differentiated whether is primary or copy recipient in the FHIR ReferralRequest resources (recipient element)

**Comment by michael-kainos** [comment]
@Kusnaditjung after investigation it seems there's no direct information on this mapping in CDS implementation guide. We need specific information how to reflect this information in FHIR. (easy to fix if we get mapping definition) I will raise this with NHS to verify

**Comment by Kusnaditjung** [comment]
There is a little description on NPFIT-FNT-TO-DPM-0808.03 Representing people and places using NHS CFH templates v1.0.pdf pages 31. (Primary and Tracker), and need to see the tabular view of the specs for [tracker](https://data.developer.nhs.uk/dms/IUC-3.0-RC1/Domains/IntegratedUrgentCare/Tabular%20View/POCD_HD200001GB02-NoEdit.html#ClinicalDocument/tracker) 
 and[ primary](https://data.developer.nhs.uk/dms/IUC-3.0-RC1/Domains/IntegratedUrgentCare/Tabular%20View/POCD_HD200001GB02-NoEdit.html#ClinicalDocument/informationRecipient). Agree this is not very obvious

**Comment by davidwaters33** [comment]
This can be derived from the participant in the 111 Report where the typeCode is 'REFT'.  If the participant is populated with the GP practice, then this is 'FOR ACTION', otherwise this is 'FOR INFORMATION'

In the Encounter Report, this recipient should have been mapped to ReferralRequest.recipient as a HealthcareService

**Comment by Kusnaditjung** [comment]
In the example given (test script in the adapter). the only participant for the CDA document is this, which seems to be from secondary (and not primary), I see from the specs.. here is below typecode from different role:

**Comment by Kusnaditjung** [comment]
Ok, I see.. at the moment there is still a number pending issue with this
(1)  the participant role is not mapped to ReferrralRequest.recipients at moment (there are only reserved for information recipient role (PRCP, and TRC).
(2)  Another thing, the recipient only contains a reference to org and does not differentiate the role PRCP, TRC or REFT. 
(3) I thought that whether a primary or copy that should already be handled in the ITK envelope from the sender side, should not be easier to output the ITK envelope somewhere (e.g FHIR message header), I have seen some commit in the other branches that look like they are going to do this.

**Comment by Kusnaditjung** [comment]
NB: @davidwaters33 I just wonder if you can get me the new transformation stylesheet  (primary and copy) that is mentioned in the new Urgent Care technical documentation. That will be useful for me to see what is to be expected in the UI Rendering. At the moment, I only have the old transformation stylesheet (for generic CDA) which I assume most probable out of date.

**Comment by davidwaters33** [comment]
> the participant role is not mapped to ReferrralRequest.recipients at moment (there are only reserved for information recipient role (PRCP, and TRC), the participant role is currently mapped to Encounter.participant of type REFT (version 0.1.0), but the mapping does not output the org
Having reviewed, it's been determined that this is done by using the `ParticipantRole` resource (https://fhir.hl7.org.uk/STU3/StructureDefinition/CareConnect-PractitionerRole-1).  The organisation information from the 111 Report is mapped to an Organisation resource in FHIR, and this is linked to the recipient through the ParticipantRole resource

**Comment by davidwaters33** [comment]
[IntegratedUrgentCare_CDA_Renderer_PrimaryRecipients.zip](https://github.com/nhsconnect/integration-adaptor-111/files/5459300/IntegratedUrgentCare_CDA_Renderer_PrimaryRecipients.zip)
Updated renderer

**Comment by Kusnaditjung** [comment]
@davidwaters33  thanks for the renderer

**Comment by kategill** [comment]
The report will have a different interaction id - primary care GPs  will receive one of the following:

urn:nhs-itk:interaction:primaryGeneralPractitionerRecipientNHS111CDADocument-v2-0
urn:nhs-itk:interaction:copyRecipientNHS111CDADocument-v2-0

Also your example in the mapping appears to be using examples from this: https://data.developer.nhs.uk/dms/IUC-3.0-RC1/Examples/examples.html

This is a release of ITK 111 DMS that was never implemented  - you need to be looking at V2.1 RC1 - this is available by Trud.

The recipient is also in the ITK header but is not the reply to field.  Always sent as follows:
<addresslist>
<address uri="urn:nhs-uk:addressing:ods:20000729"/>
<address type="2.16.840.1.113883.2.1.3.2.4.18.44" uri="2000072936"/>
</addresslist>

First is the ODS code - for most practice systems this is the correct ODS code, however it is not reliable.  In reality all GP systems should have a mapping of DOS service if - which is the second code in here - to the practice system that should receive this message.  Note there can be multiple DOS service IDs for each instance - e.g. for a practice hub or a branch site etc.

If you want a review of the SOAP ITK headers then I am happy to support.  But also you will need to review mappings as you are mapping from a spec that has not been implemented.

**Comment by Kusnaditjung** [comment]
Thanks, @kategill, I am thinking the above reply may suit better for issue #117, now that issue has been resolved in 0.3.0, so not a problem anymore. Also, the issue with the recipient address list has been handled as well in 0.3.0 (thanks for the explanation above).  This issue is particularly more about primary and recipient as represented in the CDA itself (e.g not in the ITK distribution list, which you have explained and was raised in #117), and the above issue was raised because the primary and copy recipient are output if the rendering stylesheet was used, which I have used to test end-to-end to test the sufficiency of the mapping. However, there was another reply that mentioned the adapter project does not focus on the rendering part, and that part was up to the supplier. So in the light of this, this issue becomes secondary and it is up to the kainos what take this as an issue or closed this.

**Comment by Kusnaditjung** [comment]
@kategill  About the version, I am not clear who you refer to, from the start, I used the spec from TRUD (version Release 2.0 RC 1, as this was the version mentioned when we have a meeting with David waters) (I am from test partner side), but later on, I notice that there was another similar spec on the NHS website and I think @michael-kainos  also mentioned them in one of his replies. For most issues I raised on GitHub  I have checked the RIM model in both DMS (the TRUD and NHS), they are the same so far as the message type is the same as well (POCD_MT200001GB02). I am pretty sure there will other differences, but I have not looked in full detail at the other spec pages in the NHS website yet.
